src/module.c

```diff
@@ -6463,25 +6463,27 @@ const char *RM_CallReplyProto(RedisModuleCallReply *reply, size_t *len) {
 
 /* Redis only exposes "side effects" of command execution at the end of an execution unit.
  *
- * Side effects include propagating replication/AOF operations and executing the commands of clients blocked on keys
- * modified by commands
+ * Side effects currently handled are
+ * - propagating replication/AOF operations
+ * - executing the commands of clients blocked on keys modified by commands
  *
  * In general, for modules, execution unit start and end with the taking/release of the GIL.  For module code executed
  * on the Redis main thread, this is done implicitly by Redis itself, while for module code executed on a separate
  * thread, this is done explicitly by the module itself.
  *
- * This function enables modules to flush the execution unit, thereby having side effects become visible during
- * execution.
+ * This function enables modules to flush the execution unit within the middle of an execution unit, thereby having
+ * side effects become visible and impact future operations that occur within the execution unit.
  *
- * Examples of usage might be that one is running multiple commands, one after another, with RM_Call and wants
- * processCommand() like semantics, where unblocked clients will be executed before the next RM_Call is made.
+ * Usage Example: a module runs multiple commands via RM_Call within a single execution unit, but wants
+ * processCommand() like semantics, where blocked clients that could be unblocked by a previous RM_Call operation will
+ * be executed before the next RM_Call is made.  By calling RM_FlushExecutionUnit between RM_Call operations, the
+ * module will get those semantics.
  *
- * Note: while one Modules that calls another module via RM_Call won't nest if they are operating on a separate thread,
- * as each will need to grab the GIL independently, they can nest if both operate on the main thread.  In that case,
- * it doesn't make sense for the further nested module to be able to cause side effects to the top level module, and
- * therefore this function only succeeds if running at an execution_nesting level of 1.
  */
-int RM_FlushExecutionUnit() {
+int RM_FlushExecutionUnit(RedisModuleCtx *ctx, unsigned int flags) {
```

Comment:
sadly for that to be useful we need to do a bit more.
first we need to document that flags argument (saying it's for future use).
secondly, i think we need to declare some DEFAULT constant, e.g. like we have for `REDISMODULE_BLOCK_UNBLOCK_DEFAULT`.
and then we need to decide and document how we handle future backwards compatibility.
i.e. are we gonna error on unsupported flags (in which case we should now error when 0 is used)?
or we can go with something similar to `RM_GetModuleOptionsAll()`.
i think just erroring is enough and is simpler.
@MeirShpilraien @guybe7 WDYT?

Reply:
added something like this, see if it was what you were thinking

Reply:
Thoughts about having an even more explicit API, such as: `UnblockPendingClients`, that will explicitly unblock and handle pending clients (within the same execution unit).

Reply:
@madolson can you elaborate? i don't understand what you mean, or maybe i have the wrong context.

Reply:
so it started off as similar to that (see my calling it of RM_HandleUnblockedClients() above), but was pushed in a more generic direction. but whatever naming people want is fine.

Reply:
@oranagra I suppose two things come to mind. The first is that "flush execution unit" is extremely generic, it's hard to reason about what that it should or will do as we add features to Redis. The second is I don't want to expose the concept of an execution unit to modules, since that is already a hard of enough thought to understand for us internally. Having smaller APIs that are clearer to understand seem better than one big generic one. 

Sorry for not reading the entire chain, I saw the conversation about not overloading RM_Call, which I agree with.

Reply:
@madolson i think i'd like to make it generic (i.e. execution unit) on purpose, and not start exposing RM_HandleUnblockedClients, RM_FlushPropagationQueue, etc (which are a little bit low level).
instead, i want the module to tell redis: "i'm done" and redis should do what it needs to.
this way, when in the future we add more tasks to be performed between commands, we can add them to that API, and the module / API won't need any change.

I did add a flags input argument so that if in the future we'll wanna let modules trigger just part of the operations, and skip others, we'll be able to do that.
but the default should be to run them all..

Reply:
If we want to tell Redis, "I'm done", do we want to expose the ability for modules to more clearly enter and exit execution units? Either with a `RM_ExecuteSubContext(ctx, contextFn)` which executes the contextFn in a separate execution unit. Not sure I really like this suggested API though.

Maybe I'd be okay if we renamed the API to just be something like `RM_FlushContext(ctx)` which retires the current context and returns a new one. It seems easier to reason about it in terms of context and not the underlying execution units (or maybe it does for me at least). Keeping the flags seems OK.

Reply:
i don't think i like to retire the context (good change the user will keep using it after).
also, the context is used for many other things (not just RM_Call and RM_Replicate).
I do think there's no need for RM_BegingXxxx and RM_EndXxxx since one RM_FlushXxxx can do it, and i do think we want just one generic Flush function with flags that can be extended rather than an API for each thing we need to flush.
@yossigo maybe you have some better ideas.

Reply:
> also, the context is used for many other things (not just RM_Call and RM_Replicate).

You know the API and usages better than I do, but it seems better to keep APIs constrained to what we expose today. My main fear is we box ourselves into an API contract we have to maintain. Contexts are at least exposed today, and have semi-well defined lifetimes. 

> i don't think i like to retire the context (good chance the user will keep using it after).

If only we had moved semantics. There is definitely some hints we can give to users about misuse. This API maybe needs to return an allocated Context instead of a stack allocated one, so that we can fully NULL it out.